### PR TITLE
new resource `azurerm_elastic_san_volume_group`

### PR DESCRIPTION
**Pull request by teowa**
ref:
1. [Azure doc](https://learn.microsoft.com/en-us/azure/storage/elastic-san/elastic-san-create?tabs=azure-cli#create-volume-groups)
2. [REST API specs](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json)

### CODE REVIEW

**Comment by teowa** [comment]
image

File: website/docs/r/elastic_san_volume_group.html.markdown [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+* `current_versioned_key_id` - Current Versioned Key ID.
+
+* `last_key_rotation_timestamp` - Last Key Rotation Timestamp.
+
+---
+
+An `identity` block exports the following arguments:
+
+* `principal_id` - The Principal ID for the System-Assigned Managed Identity assigned to this Elastic SAN Volume Group.
```
Comment (ms-zhenhua) [inline website/docs/r/elastic_san_volume_group.html.markdown:188]:
```suggestion
* `principal_id` - The Principal ID associated with the Managed Service Identity assigned to this Elastic SAN Volume Group.
```

File: website/docs/r/elastic_san_volume_group.html.markdown [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+* `tenant_id` - The Tenant ID for the System-Assigned Managed Identity assigned to this Elastic SAN Volume Group.
```
Comment (ms-zhenhua) [inline website/docs/r/elastic_san_volume_group.html.markdown:190]:
```suggestion
* `tenant_id` - The Tenant ID associated with this Managed Service Identity assigned to this Elastic SAN Volume Group.
```

File: website/docs/r/elastic_san_volume_group.html.markdown [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+* `id` - The ID of the Elastic SAN Volume Group.
+
+---
+
+An `encryption` block exports the following arguments:
+
+* `current_versioned_key_expiration_timestamp` - Current Versioned Key Expiration Timestamp of the Elastic SAN Volume Group.
+
+* `current_versioned_key_id` - Current Versioned Key ID.
```
Comment (ms-zhenhua) [inline website/docs/r/elastic_san_volume_group.html.markdown:180]:
```suggestion
* `current_versioned_key_id` - The ID of the current versioned Key Vault Key in use.
```

File: website/docs/r/elastic_san_volume_group.html.markdown [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+* `last_key_rotation_timestamp` - Last Key Rotation Timestamp.
```
Comment (ms-zhenhua) [inline website/docs/r/elastic_san_volume_group.html.markdown:182]:
```suggestion
* `last_key_rotation_timestamp` - The timestamp of the last rotation of the Key Vault Key.
```

File: website/docs/r/elastic_san_volume_group.html.markdown [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+In addition to the Arguments listed above - the following Attributes are exported:
+
+* `id` - The ID of the Elastic SAN Volume Group.
+
+---
+
+An `encryption` block exports the following arguments:
+
+* `current_versioned_key_expiration_timestamp` - Current Versioned Key Expiration Timestamp of the Elastic SAN Volume Group.
```
Comment (ms-zhenhua) [inline website/docs/r/elastic_san_volume_group.html.markdown:178]:
```suggestion
* `current_versioned_key_expiration_timestamp` - The timestamp of the expiration time for the current version of the customer managed key.
```

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]
Hey @teowa, 
Thanks for this PR - I've taken a look through and left some comments inline. If we can fix those up, this should be good to go 👍

Thanks!

File: internal/services/elasticsan/elastic_san_volume_group_resource.go [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+			LastKeyRotationTimestamp:               pointer.From(input.KeyVaultProperties.LastKeyRotationTimestamp),
+		},
+	}, nil
+}
+
+func ExpandVolumeGroupNetworkRules(input []ElasticSANVolumeGroupResourceNetworkRuleModel) *volumegroups.NetworkRuleSet {
+	if len(input) == 0 {
+		return &volumegroups.NetworkRuleSet{
+			VirtualNetworkRules: &[]volumegroups.VirtualNetworkRule{},
+		}
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/elasticsan/elastic_san_volume_group_resource.go:433]: if return nil, the Network Rules will not be removed during update

File: internal/services/elasticsan/elastic_san_volume_group_resource.go [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+	}
+}
+
+func FlattenVolumeGroupNetworkRules(input *volumegroups.NetworkRuleSet) []ElasticSANVolumeGroupResourceNetworkRuleModel {
+	if input == nil {
+		return []ElasticSANVolumeGroupResourceNetworkRuleModel{}
+	}
+
+	var networkRules []ElasticSANVolumeGroupResourceNetworkRuleModel
+	for _, rule := range *input.VirtualNetworkRules {
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/elasticsan/elastic_san_volume_group_resource.go:455]: will add a nil check for this

**Comment by teowa** [comment]
image

File: internal/services/elasticsan/elastic_san_volume_group_resource.go [hunk 8]
```diff
@@ -140,7 +135,7 @@ func (r ElasticSANVolumeGroupResource) Arguments() map[string]*pluginsdk.Schema
 			Type:     pluginsdk.TypeString,
 			Optional: true,
 			ValidateFunc: validation.StringInSlice([]string{
-				// None is not exposed
+				// None is not exposed, Iscsi is only allowed protocolType to connect to volume group
```
Comment (ms-zhenhua) [inline internal/services/elasticsan/elastic_san_volume_group_resource.go:138]:
```suggestion
				// None is not a valid value and service team will consider removing it in future versions.
```

File: website/docs/r/elastic_san_volume_group.html.markdown [hunk 9]
```diff
@@ -139,7 +139,7 @@ The following arguments are supported:
 
 * `network_rule` - (Optional) One or more `network_rule` blocks as defined below.
 
-* `protocol_type` - (Optional) Specifies the type of the storage target. The only possible value is `Iscsi`. Defaults to `Iscsi`.
+* `protocol_type` - (Optional) Specifies the type of the storage target. The only possible values are `Iscsi`. Defaults to `Iscsi`.
```
Comment (ms-zhenhua) [inline website/docs/r/elastic_san_volume_group.html.markdown:142]:
```suggestion
* `protocol_type` - (Optional) Specifies the type of the storage target. The only possible value is `Iscsi`. Defaults to `Iscsi`.
```

**Review by ms-zhenhua — APPROVED** [review]

**Comment by ms-zhenhua** [comment]
Hey @teowa, thank you for your updates. LGTM~

File: internal/services/elasticsan/elastic_san_volume_group_resource.go [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+}
+
+func FlattenVolumeGroupEncryption(input *volumegroups.EncryptionProperties) ([]ElasticSANVolumeGroupResourceEncryptionModel, error) {
+	if input == nil {
+		return []ElasticSANVolumeGroupResourceEncryptionModel{}, nil
+	}
+
+	var keyVaultKeyId string
+	if kv := input.KeyVaultProperties; kv != nil {
+		id, err := keyVaultParse.NewNestedItemID(pointer.From(kv.KeyVaultUri), keyVaultParse.NestedItemTypeKey, pointer.From(kv.KeyName), pointer.From(kv.KeyVersion))
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/elasticsan/elastic_san_volume_group_resource.go:395]: From [doc](https://learn.microsoft.com/en-us/azure/storage/elastic-san/elastic-san-configure-customer-managed-keys?tabs=azure-powershell%2Cnew-vg#automatic-key-version-rotation), If `key_vault_id` has no version, the backend will periodically retrieve the latest version, if version specified, we need to manually update version if key is rotated.

File: internal/services/elasticsan/elastic_san_volume_group_resource.go [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+	if len(input) == 0 {
+		return nil, nil
+	}
+
+	nestedItemId, err := keyVaultParse.ParseOptionallyVersionedNestedItemID(input[0].KeyVaultKeyId)
+	if err != nil {
+		return nil, err
+	}
+
+	result := volumegroups.EncryptionProperties{
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/elasticsan/elastic_san_volume_group_resource.go:371]: If `version` is specified, specific versioned key will be used;
if `version` is not specified, 
> Elastic SAN checks the key vault daily for a new version of the key.

File: internal/services/elasticsan/elastic_san_volume_group_resource.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/elasticsan/elastic_san_volume_group_resource.go:399]: Yes, by my test, `KeyVersion` will not be changed, another versioned Key ID is exported through `currentVersionedKeyId` field of REST API.

File: internal/services/elasticsan/elastic_san_volume_group_resource.go [hunk 12]
```diff
@@ -0,0 +1,459 @@
...
+			}
+
+			return nil
+		},
+	}
+}
+
+func ExpandVolumeGroupEncryption(input []ElasticSANVolumeGroupResourceEncryptionModel) (*volumegroups.EncryptionProperties, error) {
+	if len(input) == 0 {
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/elasticsan/elastic_san_volume_group_resource.go:367]: By my test, if `encryptionType` is updated from `EncryptionAtRestWithCustomerManagedKey` to  `EncryptionAtRestWithPlatformKey`, then `encryption` block will be ignored at backend.

File: internal/services/elasticsan/elastic_san_volume_group_resource.go [hunk 13]
```diff
@@ -0,0 +1,459 @@
...
+
+		userAssignedIdentityId = id.ID()
+	}
+
+	return []ElasticSANVolumeGroupResourceEncryptionModel{
+		{
+			KeyVaultKeyId:                          keyVaultKeyId,
+			UserAssignedIdentityId:                 userAssignedIdentityId,
+			CurrentVersionedKeyExpirationTimestamp: pointer.From(input.KeyVaultProperties.CurrentVersionedKeyExpirationTimestamp),
+			CurrentVersionedKeyId:                  pointer.From(input.KeyVaultProperties.CurrentVersionedKeyIdentifier),
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/elasticsan/elastic_san_volume_group_resource.go]: nil check added.

**Review by ms-zhenhua — APPROVED** [review]

**Comment by teowa** [comment]
image

**Review by stephybun — APPROVED** [review]
Thanks @teowa and @ms-zhenhua! LGTM 🦀

**Pull request merged**